README.md
# 静态代码收藏
学习好的代码，精华的代码，提高自己的编程能力。          
我想像一些厉害的高手，都是有一个自己的类库了，各种代码。        



changelog.md
## 1.1.31（2022-02-16）
修复微信小程序端，修改绑定的手机号码时表单验证不正常的问题
## 1.1.30（2022-01-26）
- 新增逻辑:调用uni-id-cf的logout接口后刷新设备信息中token的有效期
- 修复某些情况下前端执行logout没调用uniID.logout的问题
- 修复某些情况下报push_clientid未定义的问题
## 1.1.29（2022-01-25）
- 保存`uni_id_token`到`storage`改用异步方法，方便通过拦截器执行`token`更新后的操作
- 新增通过拦截器监听`uni_id_token`更新，调用云函数刷新刷新设备信息token有效期的API `renewDeviceTokenExpired`
- 删除留言板示例
- 修复图片验证码样式问题
## 1.1.28（2022-01-12）
删除list页第37行多了个`？`引起的报错
## 1.1.27（2022-01-11）
更新uni-id-cf为:1.0.10版，修复 限制只有 admin 用户可管理用户标签（不支持非 admin 用户操作managerMultiTag 接口）
## 1.1.26（2021-12-29）
- 性能优化，list页面使用`getTemp`，[详情](https://uniapp.dcloud.io/uniCloud/unicloud-db?id=collection)
- 拉齐uni-starter和uni-admin的schema新增：uni-id-tag.schema.json，更新：opendb-verify-codes.schema.json
- 修复首次登陆,用户id没存储到storage的问题
- 新增：执行退出登陆后，通过云函数调用`uniID.logout`
## 1.1.25（2021-12-09）
修复H5端在about页面，返回触发`uniShare.hide()`引发报错
## 1.1.24（2021-11-29）
- 新增注销用户账号的功能
- 修复在某些情况下，签到不连续7天，也获得60积分的问题
## 1.1.23（2021-11-20）
- 使用`uni.getUniverifyManager`优化一键登陆中，点击第三方登陆的逻辑：未勾选隐私政策时，toast提醒并阻止了一键登陆界面的close
- 新增支持看激励视频广告签到
## 1.1.22（2021-11-10）
删除`common/openApp.js`中可选链操作符，解决vue3版本在hbuilderX内置浏览器不兼容的问题
## 1.1.21（2021-11-10）
新增app端列表页面使用原生list下拉刷新
## 1.1.20（2021-11-08）
修复vue3版某些情况下i18n报错的问题
## 1.1.19（2021-11-08）
配置文件`uni-starter.config.js`默认关闭i18n多语言国际化
## 1.1.18（2021-10-14）
使用2.0版`uni-share`。当显示分享窗口时，监听返回操作(如：物理返回，全面屏手机侧滑)关闭分享窗口，而不是关闭当前页面。
## 1.1.17（2021-10-12）
- 更新文档
- 修复list页面where条件中缺少&符，导致的错误
## 1.1.16（2021-10-05）
在控制台提示：开启多语言国际化，将获取i18n中配置的tabbar的值覆盖pages.json中配置的tabbar的值
## 1.1.15（2021-10-02）
新增，支持配置是否开启i18n多语言国际化。
配置文件：`uni-starter.config.js`
`
"i18n":{
	"enable":true //默认启用，国际化。如果你不想使用国际化相关功能，请改为false
}
`
## 1.1.14（2021-09-30）
1. 通过微信小程序登录自动保存`sessionKey`到`uni-id-users`表
2. 我的-设置-个人资料 点击绑定手机号码，完善账号信息支持以下三种策略：
	- APP端，（如果支持）使用"通过运营商一键获取手机号码"
	- 微信小程序端，支持"一键获取微信绑定的手机号"
	- 其他端，通过手机验证码
## 1.1.13（2021-09-29）
修复search页面因多语言国际化导致的白屏问题
## 1.1.12（2021-09-28）
1. 改造微信登录逻辑，直接使用`uni.login`参数`"onlyAuthorize":true`实现
2. 修复，一键登录弹出层，已勾选“同意隐私政策协议”点击自定义登录按钮，报“你未同意隐私政策协议”的bug
## 1.1.11（2021-09-24）
优化邀请下载app页（`pages/ucenter/invite`）下载按钮闪烁的问题
## 1.1.10（2021-09-23）
修复获取验证码按钮的文字，在中文模式下显示为英文的问题
## 1.1.9（2021-09-16）
修复由多语言切换功能引起的隐私政策协议标题链接被重写的问题
## 1.1.8（2021-09-15）
更新数据表guestbook的schema中更新权限的配置
## 1.1.7（2021-09-14）
更新数据表opendb-news-articles的schema中的权限配置
## 1.1.6（2021-09-13）
纠正错误schema权限表达式`doc.uid`为`doc.user_id`
## 1.1.5（2021-09-01）
为了更直观理解路由拦截。移除路由拦截器中，默认过滤登录相关页面拦截的逻辑。确保所有白名单页面均在配置文件router.visitor中体现
## 1.1.4（2021-08-31）
修改错误的文章表`SChema`的读权限表达式
## 1.1.3（2021-08-31）
修复在微信小程序端默认语言为英文的问题
## 1.1.2（2021-08-30）
修复在微信小程序下切换语言报`locale`不存在的问题
## 1.1.1（2021-08-30）
- 解决3.2.6以下版本hbuilderx，编译的项目报`uni.setLocale`不存在的问题
## 1.1.0（2021-08-27）
- APP端支持vue3 （hbuilderx 3.2.5+）
- 支持国际化 中英文切换
- 新增留言板示例
- 修复签到的时区问题
## 1.0.48（2021-08-10）
- 修复登录成功后响应体包含`userInfo.password`的问题
- 修改了`uni-id-users`表的schema中字段username的编辑权限，防止用户通过clientDB绕过用户名不能重复的规则更新用户名的问题
## 1.0.47（2021-08-09）
- 更新文档快速体验部署流程
- 修复一键登录优先时报变量找不到的问题
## 1.0.46（2021-08-05）
清理多余文件
## 1.0.45（2021-08-05）
默认首页为nvue页面+fast
## 1.0.44（2021-08-05）
解决首页为非nvue页面时白屏的问题。
- 注意：本次在`common/appInit.js`中修改了路由拦截的逻辑，是个兼容方案；当首页为非nvue页面，路由拦截器逻辑会在加载首页时执行。接下来新版本的hx编译的uni-app项目无论首页是否为nvue都不走拦截器，保持各端逻辑一致。
## 1.0.43（2021-08-02）
1. 微信小程序端，新增：微信登录成功后，弹出是否"获取微信头像和昵称，完善个人资料"的弹框
2. APP端，新增逻辑：微信登录成功后，自动获取用户的微信昵称和头像完善用户个人资料
- 提示：因为微信的头像一旦更换，微信返回的头像url会失效。所以，以上两示例功能将url（客户端：下载到临时目录/服务端：转为Buffer）再上传到uniCloud云存储中再使用。
## 1.0.42（2021-07-29）
新增绑定手机号码页面前端校验
## 1.0.41（2021-07-27）
1. 支持vue3.0
2. 去掉App.vue全局样式，避免与非flex布局的页面样式冲突
## 1.0.40（2021-07-22）
1. 调整使用正则表达式配置强制登录功能的写法，解决在小程序端的兼容问题。
2. 新增签到功能（培养用户习惯，提升用户粘性）。支持：每日签到奖励、周期性连续7日签到，奖励翻倍。
## 1.0.39（2021-07-19）
1. 强制登录配置,新增白名单模式
2. 强制登录配置,支持正则表达式
## 1.0.38（2021-07-17）
删除多余文件
## 1.0.37（2021-07-14）
去掉配置文件：`uni-starter.config.js`，`h5` —> `url`结尾的`/`
## 1.0.36（2021-07-14）
剪切板中的邀请码，添加标识性前缀 `uniInvitationCode:`
## 1.0.35（2021-07-12）
1. H5端默认不开启，隐私权限协议签署页面。因为网页端没有什么隐私权限能被获取，目前全球仅欧盟有要求；如有需要请手动开启
2. 在列表页演示，如何在onShow生命周期获取设备位置，并在设备或者应用没有权限时自动引导。设置完毕自动重新获取。[更多点此查看插件介绍](https://ext.dcloud.net.cn/plugin?name=json-gps)
## 1.0.34（2021-07-08）
修复，打开登录页时携带参数，导致的快捷登录方式重复的问题
## 1.0.33（2021-07-06）
修复，点击短信验证码登录打开的页面不正确的问题
## 1.0.32（2021-07-06）
修复，仅配置一种快捷登录时的错误
## 1.0.31（2021-07-02）
优化项目文档
## 1.0.30（2021-07-01）
1. 简化宫格页面写法，方便理解如何控制不同状态角色的用户是否可见哪些元素。
2. uni-id-cf发送短信验证码api，默认注释掉：虚拟发送短信验证码的代码块。
3. uni-id-cf统一action名称为驼峰法
## 1.0.29（2021-06-29）
1. 修复在安卓10以下设备,操作登录获取不到oaid会直接导致登录失败的bug
2. 修复uniCloud版本为阿里云版时删除头像设置失败，腾讯云版删除头像后二次上传失败的问题
## 1.0.28（2021-06-28）
修复云函数uni-id-cf的resetPwdBySmsCode接口，未注册过的用户也能调用的问题
## 1.0.27（2021-06-25）
修改文档，新增h5版演示示例
## 1.0.26（2021-06-24）
升级用户头像上传的裁切功能，app端为原生裁剪其他端保持原来方式。数据表字段改用avatar_file存储file对象方便做图片的回显
## 1.0.25（2021-06-23）
预置uniCloud admin依赖的uniCloud文件，方便uniCloud admin与uni-starter配套使用时免做文件迁移
## 1.0.24（2021-06-23）
删除callFunction拦截器中多余的代码
## 1.0.23（2021-06-22）
更正调试遗留的uni-config-center/uni-id/config.json的tokenExpiresIn=1配置问题，改为默认值7200
## 1.0.22（2021-06-22）
1. 新增一键登录授权界面的其他快捷登录按钮
2. 优化uni-quick-login组件代码 
3. 调整隐私政策协议框勾选逻辑：在登录页面已勾选，同步勾选。如果没勾选需要手动勾选（为符合应用市场上架要求）
4. 调整登录页隐私政策协议框位置。 
5. 增强路由拦截，新增判断token是否过期。
## 1.0.21（2021-06-21）
优化：uni_modules模式使用uni-id-cf，方便uni-starter与uniCloud-admin的uni-id-cf同步更新。
## 1.0.20（2021-06-18）
1.H5端新增，强制要求用户同意隐私协议 2.兼容ios端自动设置打开下载页用户的剪切板为邀请者的inviteCode 3.成功注册用户，且请求体含邀请码inviteCode自动关联裂变关系
## 1.0.19（2021-06-17）
1.新增获取邀请码接口getUserInviteCode 2.在邀请用户下载应用页面，自动设置被邀请用户的剪切板为邀请者的code(仅支持安卓端) 3.在注册或登录并注册请求时自动添加剪切板中的请求参数 4.统一接口名称为驼峰法
## 1.0.18（2021-06-15）
修复，APP端有安装微信客户端但未显示微信登录快捷键的问题
## 1.0.17（2021-06-09）
修复，非APP端deviceInfo为空引起的登录失败问题
## 1.0.16（2021-06-08）
新增，操作注册/登录操作自动获取客户端设备：push_clientid、imei、oaid、idfa新增/更新到数据表uni-id-device新增，操作注册/登录操作自动获取客户端设备：push_clientid、imei、oaid、idfa新增/更新到数据表uni-id-device
## 1.0.15（2021-06-07）

PHP/functions.php
<?php
/**
 * 数组转换字符串（以逗号隔开）
 * @param 
 * @author Michael_xu
 * @return 
 */
function arrayToString($array)
{
    if (!is_array($array)) {
        $data_arr[] = $array;
    } else {
    	$data_arr = $array;
    }
    $data_arr = array_filter($data_arr); //数组去空
    $data_arr = array_unique($data_arr); //数组去重
    $data_arr = array_merge($data_arr);
    $string = $data_arr ? ','.implode(',', $data_arr).',' : '';
    return $string ? : '';
}

/**
 * 字符串转换数组（以逗号隔开）
 * @param 
 * @author Michael_xu
 * @return 
 */
function stringToArray($string)
{
    if (is_array($string)) {
        $data_arr = array_unique(array_filter($string));
    } else {
        $data_arr = $string ? array_unique(array_filter(explode(',', $string))) : [];
    }
    $data_arr = $data_arr ? array_merge($data_arr) : [];
    return $data_arr ? : [];
}

/**
 * 获取 IP  地理位置
 * 新浪IP 淘宝IP接口
 * @Return: array
 */
function getCity($ip = '')
{
    if($ip == ''){
        $url = "http://int.dpool.sina.com.cn/iplookup/iplookup.php?format=json";
        $ip=json_decode(file_get_contents($url),true);
        $data = $ip;
    }else{
        $url="http://ip.taobao.com/service/getIpInfo.php?ip=".$ip;
        $ip=json_decode(file_get_contents($url));   
        if((string)$ip->code=='1'){
           return false;
        }
        $data = (array)$ip->data;
    }
    
    return $data;   
}

/**
*判断浏览器类型
**/
function getBrowser() {
    $user_OSagent = $_SERVER['HTTP_USER_AGENT'];
    if (strpos($user_OSagent, "Maxthon") && strpos($user_OSagent, "MSIE")) {
        $visitor_browser = "Maxthon(Microsoft IE)";
    } elseif (strpos($user_OSagent, "Maxthon 2.0")) {
        $visitor_browser = "Maxthon 2.0";
    } elseif (strpos($user_OSagent, "Maxthon")) {
        $visitor_browser = "Maxthon";
    } elseif (strpos($user_OSagent, "Edge")) {
        $visitor_browser = "Edge";
    } elseif (strpos($user_OSagent, "Trident")) {
        $visitor_browser = "IE";
    } elseif (strpos($user_OSagent, "MSIE")) {
        $visitor_browser = "IE";
    } elseif (strpos($user_OSagent, "MSIE")) {
        $visitor_browser = "MSIE 较高版本";
    } elseif (strpos($user_OSagent, "NetCaptor")) {
        $visitor_browser = "NetCaptor";
    } elseif (strpos($user_OSagent, "Netscape")) {
        $visitor_browser = "Netscape";
    } elseif (strpos($user_OSagent, "Chrome")) {
        $visitor_browser = "Chrome";
    } elseif (strpos($user_OSagent, "Lynx")) {
        $visitor_browser = "Lynx";
    } elseif (strpos($user_OSagent, "Opera")) {
        $visitor_browser = "Opera";
    } elseif (strpos($user_OSagent, "MicroMessenger")) {
        $visitor_browser = "微信浏览器";
    } elseif (strpos($user_OSagent, "Konqueror")) {
        $visitor_browser = "Konqueror";
    } elseif (strpos($user_OSagent, "Mozilla/5.0")) {
        $visitor_browser = "Mozilla";
    } elseif (strpos($user_OSagent, "Firefox")) {
        $visitor_browser = "Firefox";
    } elseif (strpos($user_OSagent, "U")) {
        $visitor_browser = "Firefox";
    } else {
        $visitor_browser = "其它";
    }
    return $visitor_browser;
}


/**
*判断是否是移动端
**/
function isMobile()
{
    // 如果有HTTP_X_WAP_PROFILE则一定是移动设备
    if (isset ($_SERVER['HTTP_X_WAP_PROFILE']))
    {
        return true;
    }
    // 如果via信息含有wap则一定是移动设备,部分服务商会屏蔽该信息
    if (isset ($_SERVER['HTTP_VIA']))
    {
        // 找不到为flase,否则为true
        return stristr($_SERVER['HTTP_VIA'], "wap") ? true : false;
    }
    // 判断手机发送的客户端标志,兼容性有待提高,把常见的类型放到前面
    if (isset ($_SERVER['HTTP_USER_AGENT']))
    {
        $clientkeywords = array (
            'android',
            'iphone',
            'samsung',
            'ucweb',
            'wap',
            'mobile',
            'nokia',
            'huawei',
            'sony',
            'ericsson',
            'mot',
            'htc',
            'sgh',
            'lg',
            'sharp',
            'sie-',
            'philips',
            'panasonic',
            'alcatel',
            'lenovo',
            'ipod',
            'blackberry',
            'meizu',
            'netfront',
            'symbian',
            'windowsce',
            'palm',
            'operamini',
            'operamobi',
            'openwave',
            'nexusone',
            'cldc',
            'midp'
        );
        // 从HTTP_USER_AGENT中查找手机浏览器的关键字
        if (preg_match("/(" . implode('|', $clientkeywords) . ")/i", strtolower($_SERVER['HTTP_USER_AGENT'])))
        {
            return true;
        }
    }
    // 协议法，因为有可能不准确，放到最后判断
    if (isset ($_SERVER['HTTP_ACCEPT']))
    {
        // 如果只支持wml并且不支持html那一定是移动设备
        // 如果支持wml和html但是wml在html之前则是移动设备
        if ((strpos($_SERVER['HTTP_ACCEPT'], 'vnd.wap.wml') !== false) && (strpos($_SERVER['HTTP_ACCEPT'], 'text/html') === false || (strpos($_SERVER['HTTP_ACCEPT'], 'vnd.wap.wml') < strpos($_SERVER['HTTP_ACCEPT'], 'text/html'))))
        {
            return true;
        }
    }
    return false;
}

/**
 * 是否为移动端
 */
function is_mobile()
{
    if (isset($_SERVER['HTTP_X_WAP_PROFILE'])) {
        return true;
    }
    if (isset($_SERVER['HTTP_VIA'])) {
        return stristr($_SERVER['HTTP_VIA'], "wap") ? true : false;
    }
    if (isset($_SERVER['HTTP_USER_AGENT'])) {
        $clientkeywords = array('nokia', 'sony', 'ericsson', 'mot', 'samsung', 'htc', 'sgh', 'lg', 'sharp', 'sie-', 'philips', 'panasonic', 'alcatel', 'lenovo', 'iphone', 'ipod', 'blackberry', 'meizu', 'android', 'netfront', 'symbian', 'ucweb', 'windowsce', 'palm', 'operamini', 'operamobi', 'openwave', 'nexusone', 'cldc', 'midp', 'wap', 'mobile');
        if (preg_match("/(" . implode('|', $clientkeywords) . ")/i", strtolower($_SERVER['HTTP_USER_AGENT']))) {
            return true;
        }
    }
    if (isset($_SERVER['HTTP_ACCEPT'])) {
        if ((strpos($_SERVER['HTTP_ACCEPT'], 'vnd.wap.wml') !== false) && (strpos($_SERVER['HTTP_ACCEPT'], 'textml') === false || (strpos($_SERVER['HTTP_ACCEPT'], 'vnd.wap.wml') < strpos($_SERVER['HTTP_ACCEPT'], 'textml')))) {
            return true;
        }
    }
    return false;
}



//字符串转Unicode编码
function unicode_encode($strLong) {
    $strArr = preg_split('/(?<!^)(?!$)/u', $strLong);//拆分字符串为数组(含中文字符)
    $resUnicode = '';
    foreach ($strArr as $str)
    {
        $bin_str = '';
        $arr = is_array($str) ? $str : str_split($str);//获取字符内部数组表示,此时$arr应类似array(228, 189, 160)
        foreach ($arr as $value)
        {
            $bin_str .= decbin(ord($value));//转成数字再转成二进制字符串,$bin_str应类似111001001011110110100000,如果是汉字"你"
        }
        $bin_str = preg_replace('/^.{4}(.{4}).{2}(.{6}).{2}(.{6})$/', '$1$2$3', $bin_str);//正则截取, $bin_str应类似0100111101100000,如果是汉字"你"
        $unicode = dechex(bindec($bin_str));//返回unicode十六进制
        $_sup = '';
        for ($i = 0; $i < 4 - strlen($unicode); $i++)
        {
            $_sup .= '0';//补位高字节 0
        }
        $str =  '\\u' . $_sup . $unicode; //加上 \u  返回
        $resUnicode .= $str;
    }
    return $resUnicode;
  }
  //Unicode编码转字符串方法1
  function unicode_decode($name)
  {
    // 转换编码，将Unicode编码转换成可以浏览的utf-8编码
    $pattern = '/([\w]+)|(\\\u([\w]{4}))/i';
    preg_match_all($pattern, $name, $matches);
    if (!empty($matches))
    {
      $name = '';
      for ($j = 0; $j < count($matches[0]); $j++)
      {
        $str = $matches[0][$j];
        if (strpos($str, '\\u') === 0)
        {
          $code = base_convert(substr($str, 2, 2), 16, 10);
          $code2 = base_convert(substr($str, 4), 16, 10);
          $c = chr($code).chr($code2);
          $c = iconv('UCS-2', 'UTF-8', $c);
          $name .= $c;
        }
        else
        {
          $name .= $str;
        }
      }
    }
    return $name;
  }

//Unicode编码转字符串
function unicode_decode2($str){
$json = '{"str":"' . $str . '"}';
$arr = json_decode($json, true);
if (empty($arr)) return '';
return $arr['str'];
}



/**
 * 处理提取学信网学籍信息
 */
function extract_xuexin_info($html)
{
    $dom = new DOMDocument();
    $dom->loadHTMLFile("./test.html");

    $fixedPart = $dom->getElementById("fixedPart");

    $data = [
        "username_avatar" => "",
        "sex" => "未知",
        "academy" => "",
        "major" => "",
        "join_school_time" => "",
        "student_status" => "",
    ];

    // 提取头像
    $imgs = $fixedPart->getElementsByTagName("img");
    foreach($imgs as $img) {
        if ($img->getAttribute('class') == 'by_img') {
            $data["username_avatar"] = "https://www.chsi.com.cn".$img->getAttribute('src');
        }
    }

    // 提取院校、专业、入学时间、学籍状态
    $keys = [
        "性别" => "sex",
        "院校" => "academy",
        "专业" => "major",
        "入学时间" => "join_school_time",
        "学籍状态" => "student_status",
    ];
    $xpath = new DOMXPath($dom);
    $tds = $xpath->query('//*[@id="fixedPart"]//td');
    foreach ($tds as $i => $td) {
        // 键对应的下一个td就是对应的值
        if (array_key_exists(trim($td->textContent), $keys)) {
            $key = $keys[$td->textContent];
            $data[$key] = trim($tds[$i+1]->textContent); 
        }
    }

    return $data;
}


/**
 * Convert a key => value array to a 'key: value' array for headers
 *
 * @param array $array Dictionary of header values
 * @return array List of headers
 */
function flatten($array) {
    $return = array();
    foreach ($array as $key => $value) {
        $return[] = sprintf('%s: %s', $key, $value);
    }
    return $return;
}

/**
 * User: 意象信息科技 lr
 * Desc: 下载文件
 * @param $url 文件url
 * @param $save_dir 保存目录
 * @param $file_name 文件名
 * @return string
 */
function download_file($url, $save_dir, $file_name)
{
    if (!file_exists($save_dir)) {
        mkdir($save_dir, 0775, true);
    }
    $file_src = $save_dir . $file_name;
    file_exists($file_src) && unlink($file_src);
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
    $file = curl_exec($ch);
    curl_close($ch);
    $resource = fopen($file_src, 'a');
    fwrite($resource, $file);
    fclose($resource);
    if (filesize($file_src) == 0) {
        unlink($file_src);
        return '';
    }
    return $file_src;
}

/**
 * 线性结构转换成树形结构
 * @param array $data 线性结构数组
 * @param string $sub_key_name 自动生成子数组名
 * @param string $id_name 数组id名
 * @param string $parent_id_name 数组祖先id名
 * @param int $parent_id 此值请勿给参数
 * @return array
 */
function linear_to_tree($data, $sub_key_name = 'sub', $id_name = 'id', $parent_id_name = 'pid', $parent_id = 0)
{
  $tree = [];
  foreach ($data as $row) {
    if ($row[$parent_id_name] == $parent_id) {
      $temp = $row;
      $temp[$sub_key_name] = linear_to_tree($data, $sub_key_name, $id_name, $parent_id_name, $row[$id_name]);
      $tree[] = $temp;
    }
  }
  return $tree;
}

/**
 * 生成会员码
 * @return 会员码
 */
function create_user_sn($prefix = '', $length = 8)
{
    $rand_str = '';
    for ($i = 0; $i < $length; $i++) {
        $rand_str .= mt_rand(0, 9);
    }
    $sn = $prefix . $rand_str;
    $user = User::where(['sn' => $sn])->findOrEmpty();
    if (!$user->isEmpty()) {
        return create_user_sn($prefix, $length);
    }
    return $sn;
}

//生成用户邀请码
function generate_invite_code()
{
    $letter_all = range('A', 'Z');
    shuffle($letter_all);
    //排除I、O字母
    $letter_array = array_diff($letter_all, ['I', 'O', 'D']);
    //排除1、0
    $num_array = range('2', '9');
    shuffle($num_array);

    $pattern = array_merge($num_array, $letter_array, $num_array);
    shuffle($pattern);
    $pattern = array_values($pattern);

    $code = '';
    for ($i = 0; $i < 6; $i++) {
        $code .= $pattern[mt_rand(0, count($pattern) - 1)];
    }

    $code = strtoupper($code);
    $check = User::where('distribution_code', $code)->findOrEmpty();
    if (!$check->isEmpty()) {
        return generate_invite_code();
    }
    return $code;
}

/**
 * User: 意象信息科技 lr
 * Desc：生成密码密文
 * @param $plaintext string 明文
 * @param $salt string 密码盐
 * @return string
 */
function create_password($plaintext, $salt)
{
    $salt = md5('y' . $salt . 'x');
    $salt .= '2021';
    return md5($plaintext . $salt);
}

/**
 * User: 意象信息科技 mjf
 * Desc: 用时间生成订单编号
 * @param $table
 * @param $field
 * @param string $prefix
 * @param int $rand_suffix_length
 * @param array $pool
 * @return string
 * @throws \think\db\exception\DataNotFoundException
 * @throws \think\db\exception\DbException
 * @throws \think\db\exception\ModelNotFoundException
 */
function createSn($table, $field, $prefix = '', $rand_suffix_length = 4, $pool = [])
{
    $suffix = '';
    for ($i = 0; $i < $rand_suffix_length; $i++) {
        if (empty($pool)) {
            $suffix .= rand(0, 9);
        } else {
            $suffix .= $pool[array_rand($pool)];
        }
    }
    $sn = $prefix . date('YmdHis') . $suffix;
    if (Db::name($table)->where($field, $sn)->find()) {
        return createSn($table, $field, $prefix, $rand_suffix_length, $pool);
    }
    return $sn;
}

/**
 * note 生成验证码
 * @param int $length 验证码长度
 * @return string
 */
function create_sms_code($length = 4)
{
    $code = '';
    for ($i = 0; $i < $length; $i++) {
        $code .= rand(0, 9);
    }
    return $code;
}

/**
 * 生成商品编码
 * 8位
 */
function create_goods_code($shop_id)
{
    $code =  mt_rand(10000000, 99999999);
    $goods = Goods::where([
        'code' => $code,
        'shop_id' => $shop_id,
        'del' => 0
    ])->findOrEmpty();
    if($goods->isEmpty()) {
        return $code;
    }
    create_goods_code();
}

/**
 * 是否在cli模式
 */
if (!function_exists('is_cli')) {
    function is_cli()
    {
        return preg_match("/cli/i", php_sapi_name()) ? true : false;
    }
}

/**
 * Notes:判断文件是否存在（远程和本地文件）
 * @param $file string 完整的文件链接
 * @return bool
 */
function check_file_exists($file)
{
    //远程文件
    if ('http' == strtolower(substr($file, 0, 4))) {

        $header = get_headers($file, true);

        return isset($header[0]) && (strpos($header[0], '200') || strpos($header[0], '304'));

    } else {

        return file_exists($file);
    }
}

/**
 * 将图片切成圆角
 */
function rounded_corner($src_img)
{
    $w = imagesx($src_img);//微信头像宽度 正方形的
    $h = imagesy($src_img);//微信头像宽度 正方形的
    $w = min($w, $h);
    $h = $w;
    $img = imagecreatetruecolor($w, $h);
    //这一句一定要有
    imagesavealpha($img, true);
    //拾取一个完全透明的颜色,最后一个参数127为全透明
    $bg = imagecolorallocatealpha($img, 255, 255, 255, 127);
    imagefill($img, 0, 0, $bg);
    $r = $w / 2; //圆半径
//    $y_x = $r; //圆心X坐标
//    $y_y = $r; //圆心Y坐标
    for ($x = 0; $x < $w; $x++) {
        for ($y = 0; $y < $h; $y++) {
            $rgbColor = imagecolorat($src_img, $x, $y);
            if (((($x - $r) * ($x - $r) + ($y - $r) * ($y - $r)) < ($r * $r))) {
                imagesetpixel($img, $x, $y, $rgbColor);
            }
        }
    }
    unset($src_img);
    return $img;
}

/**
 * Notes:去掉名称中的表情
 * @param $str
 * @return string|string[]|null
 * @author: cjhao 2021/3/29 15:56
 */
function filterEmoji($str)
{
    $str = preg_replace_callback(
        '/./u',
        function (array $match) {
            return strlen($match[0]) >= 4 ? '' : $match[0];
        },
        $str);
    return $str;
}

/**
 * Notes:生成一个范围内的随机浮点数
 * @param int $min 最小值
 * @param int $max 最大值
 * @return float|int 返回随机数
 */
function random_float($min = 0, $max = 1)
{
    return $min + mt_rand() / mt_getrandmax() * ($max - $min);
}

/**
 * Notes: 获取文件扩展名
 * @param $file
 * @author 段誉(2021/7/7 18:03)
 * @return mixed
 */
if (!function_exists('get_extension')) {
    function get_extension($file)
    {
        return pathinfo($file, PATHINFO_EXTENSION);
    }
}

/**
 * Notes: 删除目标目录
 * @param $path
 * @param $delDir
 * @author 段誉(2021/7/7 18:19)
 * @return bool
 */
if (!function_exists('del_target_dir')) {
    function del_target_dir($path, $delDir)
    {
        $handle = opendir($path);
        if ($handle) {
            while (false !== ($item = readdir($handle))) {
                if ($item != "." && $item != "..") {
                    if (is_dir("$path/$item")) {
                        del_target_dir("$path/$item", $delDir);
                    } else {
                        unlink("$path/$item");
                    }
                }
            }
            closedir($handle);
            if ($delDir) {
                return rmdir($path);
            }
        } else {
            if (file_exists($path)) {
                return unlink($path);
            }
            return false;
        }
    }
}


/**
 * Notes: 获取本地版本数据
 * @return mixed
 * @author 段誉(2021/7/7 18:18)
 */
if (!function_exists('local_version')) {
    function local_version()
    {
        if(!file_exists('./upgrade/')) {
            // 若文件夹不存在，先创建文件夹
            mkdir('./upgrade/', 0777, true);
        }
        if(!file_exists('./upgrade/version.json')) {
            // 获取本地版本号
            $version = config('project.version');
            $data = ['version' => $version];
            $src = './upgrade/version.json';
            // 新建文件
            file_put_contents($src, json_encode($data, JSON_UNESCAPED_UNICODE));
        }

        $json_string = file_get_contents('./upgrade/version.json');
        // 用参数true把JSON字符串强制转成PHP数组
        $data = json_decode($json_string, true);
        return $data;
    }
}

/**
 * Notes: 获取ip
 * @author 段誉(2021/7/9 10:19)
 * @return array|false|mixed|string
 */
if (!function_exists('get_client_ip')) {
    function get_client_ip()
    {
        if ($_SERVER['REMOTE_ADDR']) {
            $cip = $_SERVER['REMOTE_ADDR'];
        } elseif (getenv("REMOTE_ADDR")) {
            $cip = getenv("REMOTE_ADDR");
        } elseif (getenv("HTTP_CLIENT_IP")) {
            $cip = getenv("HTTP_CLIENT_IP");
        } else {
            $cip = "unknown";
        }
        return $cip;
    }
}

/**
 * 远程图片转换为base64编码
 * 
 * @from https://stackoverflow.com/questions/4343715/php-how-to-convert-an-image-from-url-to-base64
 * @param string $img_url
 * @return bool|string
 */
function remoteImgToBase64($img_url)
{
    $image = file_get_contents($img_url);
    if ($image !== false){
        return 'data:image/jpg;base64,'.base64_encode($image);

    }
    return false;
}


/**
 * 等同于JavaScript的 encodeURIComponent
 *
 * @param string $str
 * @return string
 */
function encodeURIComponent($str) {
    $revert = array('%21'=>'!', '%2A'=>'*', '%27'=>"'", '%28'=>'(', '%29'=>')');
    return strtr(rawurlencode($str), $revert);
}



/**
 * 获取当前页面完整URL
 * @doc https://www.cnblogs.com/xuzhengzong/p/6856367.html
 * 
 * @return void
 */
function get_current_url()
{
    $current_url='http://';
    if(isset($_SERVER['HTTPS'])&&$_SERVER['HTTPS']=='on'){
        $current_url='https://';
    }
    if($_SERVER['SERVER_PORT']!='80'){
        $current_url.=$_SERVER['SERVER_NAME'].':'.$_SERVER['SERVER_PORT'].$_SERVER['REQUEST_URI'];
    }else{
        $current_url.=$_SERVER['SERVER_NAME'].$_SERVER['REQUEST_URI'];
    }

    return $current_url;
}